### PR TITLE
Update Bundler deprecation unsupported check to use detected version

### PR DESCRIPTION
**Pull request by kbukum1**
### **What are you trying to accomplish?**

This PR updates the handling of deprecation and unsupported checks for Bundler by ensuring that `detected_version` is consistently used instead of `raw_version`. 

This change improves the accuracy of deprecation notices and resolves edge cases where `raw_version` is either `nil` or different from the `detected_version`. It also refines the notice generation and logging behavior to handle all scenarios correctly.

**Why?**
- To ensure the consistency and correctness of version checks across Dependabot's ecosystem.
- To fix potential issues with mismatched version usage in deprecation and unsupported logic.
- To enhance the reliability and clarity of notices for deprecated and unsupported versions.

---

### **Anything you want to highlight for special attention from reviewers?**

- The logic in `unsupported?` and related methods was updated to handle cases where `raw_version` is `nil`.
- Comprehensive test cases have been added to cover scenarios where:
  - `detected_version` is deprecated or unsupported.
  - `raw_version` is `nil`, deprecated, or different from `detected_version`.
- The `NoticeHelpers` module was updated and thoroughly tested to ensure proper notice generation and logging behavior.

---

### **How will you know you've accomplished your goal?**

- All tests, including the complete test suite for `NoticeHelpers` and deprecation logic, pass successfully.
- Deprecation and unsupported checks now use `detected_version` consistently, addressing any discrepancies.
- Notice generation and logging correctly reflect the updated logic, including edge cases.

---

### **Checklist**

- [x] I have run the complete test suite to ensure all tests and linters pass.
- [x] I have thoroughly tested my code changes to ensure they work as expected, including adding additional tests for new functionality.
- [x] I have written clear and descriptive commit messages.
- [x] I have provided a detailed description of the changes in the pull request, including the problem it addresses, how it fixes the problem, and any relevant details about the implementation.
- [x] I have ensured that the code is well-documented and easy to understand.

### CODE REVIEW

**Comment by kbukum1** [comment]
This PR is the fix for bundler based on the change made in the following PR.

- https://github.com/dependabot/dependabot-core/pull/11190

File: bundler/spec/dependabot/bundler/package_manager_spec.rb [hunk 1]
```diff
@@ -6,52 +6,41 @@
...
-    context "when version is a String" do
-      let(:version) { "2" }
-
-      it "sets the version correctly" do
-        expect(package_manager.version).to eq(Dependabot::Bundler::Version.new(version))
-      end
-
-      it "sets the name correctly" do
-        expect(package_manager.name).to eq(Dependabot::Bundler::PACKAGE_MANAGER)
-      end
```
Comment (jeffwidman) [inline bundler/spec/dependabot/bundler/package_manager_spec.rb:22]: I'm unclear why some of these test cases are deleted but I don't see a corresponding addition... are they no longer relevant for some reason?

File: common/spec/spec_helper.rb [hunk 2]
```diff
@@ -202,3 +202,12 @@ def command_fixture(name)
...
 end
+
+# Define an anonymous subclass of Dependabot::Requirement for testing purposes
+TestRequirement = Class.new(Dependabot::Requirement) do
+  # Initialize with comma-separated requirement constraints
+  def initialize(constraint_string)
+    requirements = constraint_string.split(",").map(&:strip)
+    super(requirements)
+  end
+end
```
Comment (jeffwidman) [inline common/spec/spec_helper.rb:213]: I'm not understanding the purpose of this and why the input would be comma-separated? Can you give an example?

I only took a quick skim of the code, but the examples I see in the tests are all ruby strings separated by periods for major.minor.patch... am I missing something?

**Review by jeffwidman — COMMENTED** [review]
The code changes look relatively straightforward, but I had some questions about the tests...

File: bundler/spec/dependabot/bundler/package_manager_spec.rb (hunk 1 above)
Comment (kbukum1) [reply bundler/spec/dependabot/bundler/package_manager_spec.rb:22]: Actually it is here?

https://github.com/dependabot/dependabot-core/pull/11222/files#diff-b094477eee8d153bf78bd61ca925d4e359d234ab4ea374b8f68ab8a46040c743R28

File: common/spec/spec_helper.rb (hunk 2 above)
Comment (kbukum1) [reply common/spec/spec_helper.rb:213]: I have encountered an error due to the reinitialization of `TestRequirement`, as it was not scoped within a module. To resolve this, I consolidated its definition into a shared module so both specs can reference it without conflict.

The changes affect the following specs:

- [[common/spec/dependabot/requirement_spec.rb](https://github.com/dependabot/dependabot-core/pull/11222/files#diff-f05a6c07bbd6c9c68dd7757eab181c1a04329a2ac3c8397ea87eb850d53e0eafL9)](https://github.com/dependabot/dependabot-core/pull/11222/files#diff-f05a6c07bbd6c9c68dd7757eab181c1a04329a2ac3c8397ea87eb850d53e0eafL9)
- [[common/spec/dependabot/ecosystem_spec.rb](https://github.com/dependabot/dependabot-core/pull/11222/files#diff-595c845a160a5abac7b6b0da97e0a1a846a09dc091efc91524354cf0a19bda9cL9)](https://github.com/dependabot/dependabot-core/pull/11222/files#diff-595c845a160a5abac7b6b0da97e0a1a846a09dc091efc91524354cf0a19bda9cL9)

By moving the definition to a shared location, we ensure both specs can use it seamlessly, avoiding reinitialization errors.

Comment (kbukum1) [reply common/spec/spec_helper.rb:213]: As you can see the bundler requirement it is also splitting by `,` (` req_string.split(",").map(&:strip)` in the following code. I think we don't want to include ecosystem requirement implementation but we do try to follow same approach to create test requirements. 

```
def initialize(*requirements)
```

```
# typed: true
# frozen_string_literal: true

require "sorbet-runtime"

require "dependabot/requirement"
require "dependabot/utils"

module Dependabot
  module Bundler
    class Requirement < Dependabot::Requirement
      extend T::Sig

      # Patches Gem::Requirement to make it accept requirement strings like
      # "~> 4.2.5, >= 4.2.5.1" without first needing to split them.
      def initialize(*requirements)
        requirements = requirements.flatten.flat_map do |req_string|
          req_string.split(",").map(&:strip)
        end

        super(requirements)
      end
    end
  end
end

Dependabot::Utils
  .register_requirement_class("bundler", Dependabot::Bundler::Requirement)
```

Comment (kbukum1) [reply common/spec/spec_helper.rb:213]: It is about seperating multiple requirements ssuch as 

`>= 1.0.1`, `<1.2.2`

Comment (jeffwidman) [reply common/spec/spec_helper.rb:213]: Oh, I get it now... that makes sense.

For some reason I thought it was trying to handle things like `version=2,1,13`... a "requirement" specification makes much more sense.

Comment (kbukum1) [reply common/spec/spec_helper.rb:213]: Thanks for checking. Actually this is already existing one, I just moved it into here.

File: bundler/spec/dependabot/bundler/package_manager_spec.rb (hunk 1 above)
Comment (kbukum1) [reply bundler/spec/dependabot/bundler/package_manager_spec.rb:22]: As discussed, we have improved our specs to incorporate the detected version, including the raw version, as part of the changes.

**Review by jeffwidman — APPROVED** [review]
Had a good call with @kbukum1 and this seems relatively straightforward to me. And if I'm missing something and it goes sideways, it shouldn't break things too badly.
